tests/test_circuit_grid.py

```diff
@@ -0,0 +1,251 @@
+
+        self.assertEqual(self.grid.selected_column, 0)
+        self.assertEqual(self.grid.selected_wire, 0)
+
+    def test_get_selected_node_gate_part(self):
+        """
+        Test getting gate on currently selected node
+        """
+
+        node1 = CircuitGridNode(node_types.Y)
+        node2 = CircuitGridNode(node_types.X)
+        node3 = CircuitGridNode(node_types.Z)
+
+        self.circuit_grid_model.set_node(0, 1, node1)
+        self.circuit_grid_model.set_node(1, 0, node2)
+        self.circuit_grid_model.set_node(2, 0, node3)
+
+        self.grid.move_to_adjacent_node(MOVE_RIGHT)
+
+        selected_gate = self.grid.get_selected_node_gate_part()
+        self.assertEqual(node1.node_type, selected_gate)
+
+        self.grid.move_to_adjacent_node(MOVE_LEFT)
+        self.grid.move_to_adjacent_node(MOVE_DOWN)
+
+        selected_gate = self.grid.get_selected_node_gate_part()
+        self.assertEqual(node2.node_type, selected_gate)
+
+        self.grid.move_to_adjacent_node(MOVE_DOWN)
+
+        selected_gate = self.grid.get_selected_node_gate_part()
+        self.assertEqual(node3.node_type, selected_gate)
+
+    def test_handle_single_qubit_gate_inputs(self):
+        """
+        Test handling input placing single
+        qubit gates on circuit grid
+        """
+
+        # 0 X-------
+        # 1 Y------
+        # 2 Z------
+        # 3 H------
+
+        self.grid.handle_input_x()
+        self.grid.move_to_adjacent_node(MOVE_DOWN)
+        self.grid.handle_input_y()
+        self.grid.move_to_adjacent_node(MOVE_DOWN)
+        self.grid.handle_input_z()
+        self.grid.move_to_adjacent_node(MOVE_DOWN)
+        self.grid.handle_input_h()
+
+        self.assertEqual(node_types.X, self.circuit_grid_model.get_node_gate_part(0, 0))
+        self.assertEqual(node_types.Y, self.circuit_grid_model.get_node_gate_part(1, 0))
+        self.assertEqual(node_types.Z, self.circuit_grid_model.get_node_gate_part(2, 0))
+        self.assertEqual(node_types.H, self.circuit_grid_model.get_node_gate_part(3, 0))
+
+        # 0 ---------
+        # 1 ---------
+        # 2 ---------
+        # 3 ---------
+
+        self.grid.handle_input_delete()
+        self.grid.move_to_adjacent_node(MOVE_UP)
+        self.grid.handle_input_delete()
+        self.grid.move_to_adjacent_node(MOVE_UP)
+        self.grid.handle_input_delete()
+        self.grid.move_to_adjacent_node(MOVE_UP)
+        self.grid.handle_input_delete()
+
+        self.assertEqual(
+            node_types.EMPTY, self.circuit_grid_model.get_node_gate_part(1, 0)
+        )
+        self.assertEqual(
+            node_types.EMPTY, self.circuit_grid_model.get_node_gate_part(1, 0)
+        )
+        self.assertEqual(
+            node_types.EMPTY, self.circuit_grid_model.get_node_gate_part(2, 0)
+        )
+        self.assertEqual(
+            node_types.EMPTY, self.circuit_grid_model.get_node_gate_part(3, 0)
+        )
+
+    def test_handle_multi_gate_inputs(self):
```

Comment:
Can you also test for cases when the controlled node is below the wire. For example, above the wire of X and Y and below the wire for Z and H.

Reply:
Y and Z are on the second wire now, and controlled from below. I did this instead as pylint was emitting a duplicate-code message.